# diff --git a/llvm/cmake/modules/AddLLVM.cmake b/llvm/cmake/modules/AddLLVM.cmake
# index 230620c37027..add525ae59b0 100644
# --- a/llvm/cmake/modules/AddLLVM.cmake
# +++ b/llvm/cmake/modules/AddLLVM.cmake
# @@ -1,3 +1,6 @@
# +# Set to an arbitrary directory to silence GNUInstallDirs warnings
# +# regarding being unable to determine libdir.
# +set(CMAKE_INSTALL_LIBDIR "lib")
#  include(GNUInstallDirs)
#  include(LLVMDistributionSupport)
#  include(LLVMProcessSources)
# @@ -2352,8 +2355,13 @@ function(llvm_setup_rpath name)
#      # FIXME: update this when there is better solution.
#      set(_install_rpath "${LLVM_LIBRARY_OUTPUT_INTDIR}" "${CMAKE_INSTALL_PREFIX}/lib${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
#    elseif(UNIX)
# -    set(_build_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
# -    set(_install_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}")
# +    set(_build_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/${LLVM_DEFAULT_TARGET_TRIPLE}"
# +    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/clang/${CLANG_VERSION}/lib/${LLVM_DEFAULT_TARGET_TRIPLE}"
# +    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}"
# +    ${extra_libdir})
# +    set(_install_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/${LLVM_DEFAULT_TARGET_TRIPLE}"
# +    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/clang/${CLANG_VERSION}/lib/${LLVM_DEFAULT_TARGET_TRIPLE}"
# +    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}")
#      if(${CMAKE_SYSTEM_NAME} MATCHES "(FreeBSD|DragonFly)")
#        set_property(TARGET ${name} APPEND_STRING PROPERTY
#                     LINK_FLAGS " -Wl,-z,origin ")


diff --git a/llvm/cmake/modules/AddLLVM.cmake b/llvm/cmake/modules/AddLLVM.cmake
index 230620c37027..43c2e6534f0d 100644
--- a/llvm/cmake/modules/AddLLVM.cmake
+++ b/llvm/cmake/modules/AddLLVM.cmake
@@ -1,8 +1,35 @@
+# Set to an arbitrary directory to silence GNUInstallDirs warnings
+# regarding being unable to determine libdir.
+set(CMAKE_INSTALL_LIBDIR "lib")
 include(GNUInstallDirs)
 include(LLVMDistributionSupport)
 include(LLVMProcessSources)
 include(LLVM-Config)
 include(DetermineGCCCompatible)
+list(APPEND CMAKE_MODULE_PATH "$ENV{MAKE_DIR}/files")
+include(env)
+
+if(CMAKE_SYSROOT_COMPILE)
+  set(_cmake_sysroot_compile "${CMAKE_SYSROOT_COMPILE}")
+else()
+  set(_cmake_sysroot_compile "${CMAKE_SYSROOT}")
+endif()
+
+if (NOT DEFINED _cmake_sysroot_compile OR _cmake_sysroot_compile STREQUAL "")
+  set (_cmake_sysroot_compile "$ENV{CMAKE_SYSROOT_COMPILE}")
+  if (NOT DEFINED _cmake_sysroot_compile OR _cmake_sysroot_compile STREQUAL "")
+    set (_cmake_sysroot_compile "$ENV{CMAKE_SYSROOT}")
+    if (NOT DEFINED _cmake_sysroot_compile OR _cmake_sysroot_compile STREQUAL "")
+      message (FATAL_ERROR "Failed to set \${_cmake_sysroot_compile} = ${_cmake_sysroot_compile}")
+    endif ()
+  endif ()
+endif ()
+
+filter_prefixed("${CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES}"   ${_cmake_sysroot_compile} CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES)
+filter_prefixed("${CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES}" ${_cmake_sysroot_compile} CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES)
+filter_prefixed("${CMAKE_ASM_IMPLICIT_INCLUDE_DIRECTORIES}" ${_cmake_sysroot_compile} CMAKE_ASM_IMPLICIT_INCLUDE_DIRECTORIES)
+
+unset(_cmake_sysroot_compile)
 
 function(llvm_update_compile_flags name)
   get_property(sources TARGET ${name} PROPERTY SOURCES)
@@ -2352,8 +2379,13 @@ function(llvm_setup_rpath name)
     # FIXME: update this when there is better solution.
     set(_install_rpath "${LLVM_LIBRARY_OUTPUT_INTDIR}" "${CMAKE_INSTALL_PREFIX}/lib${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
   elseif(UNIX)
-    set(_build_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}" ${extra_libdir})
-    set(_install_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}")
+    set(_build_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/${LLVM_DEFAULT_TARGET_TRIPLE}"
+    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/clang/${CLANG_VERSION}/lib/${LLVM_DEFAULT_TARGET_TRIPLE}"
+    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}"
+    ${extra_libdir})
+    set(_install_rpath "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/${LLVM_DEFAULT_TARGET_TRIPLE}"
+    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}/clang/${CLANG_VERSION}/lib/${LLVM_DEFAULT_TARGET_TRIPLE}"
+    "\$ORIGIN/../lib${LLVM_LIBDIR_SUFFIX}")
     if(${CMAKE_SYSTEM_NAME} MATCHES "(FreeBSD|DragonFly)")
       set_property(TARGET ${name} APPEND_STRING PROPERTY
                    LINK_FLAGS " -Wl,-z,origin ")







